pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java

```diff
@@ -757,57 +753,157 @@ private void createDerivedColumnV1Indices(String column, FunctionEvaluator funct
-              : new MultiValueUnsortedForwardIndexCreator(_indexDir, column, cardinality, numDocs,
-                  indexCreationInfo.getTotalNumberOfEntries())) {
-            for (int i = 0; i < numDocs; i++) {
-              forwardIndexCreator.putDictIdMV(dictionaryCreator.indexOfMV(outputValues[i]));
-            }
+          for (Object outputValue : outputValues) {
+            forwardIndexCreator.putDictIdMV(dictionaryCreator.indexOfMV(outputValue));
           }
         }
-
         // Add the column metadata
         SegmentColumnarIndexCreator.addColumnMetadataInfo(_segmentProperties, column, indexCreationInfo, numDocs,
             fieldSpec, true, dictionaryCreator.getNumBytesPerEntry());
       }
-    } finally {
-      for (ValueReader valueReader : valueReaders) {
-        valueReader.close();
+    }
+  }
+
+  /**
+   * Helper method to create a forward index for a raw encoded column with derived values.
+   */
+  private void createDerivedColumnForwardIndexWithoutDictionary(String column, FieldSpec fieldSpec,
+      Object[] outputValues, ColumnIndexCreationInfo indexCreationInfo)
+      throws Exception {
+
+    // Create forward index
+    int numDocs = outputValues.length;
+    boolean isSingleValue = fieldSpec.isSingleValueField();
+
+    try (ForwardIndexCreator forwardIndexCreator
+        = getForwardIndexCreator(fieldSpec, indexCreationInfo, numDocs, column, false)) {
+      if (isSingleValue) {
+        for (Object outputValue : outputValues) {
+          switch (fieldSpec.getDataType().getStoredType()) {
+            // Casts are safe here because we've already done the conversion in createDerivedColumnV1Indices
+            case INT:
+              forwardIndexCreator.putInt((int) outputValue);
+              break;
+            case LONG:
+              forwardIndexCreator.putLong((long) outputValue);
+              break;
+            case FLOAT:
+              forwardIndexCreator.putFloat((float) outputValue);
+              break;
+            case DOUBLE:
+              forwardIndexCreator.putDouble((double) outputValue);
+              break;
+            case BIG_DECIMAL:
+              forwardIndexCreator.putBigDecimal((BigDecimal) outputValue);
+              break;
+            case STRING:
+              forwardIndexCreator.putString((String) outputValue);
+              break;
+            case BYTES:
+              forwardIndexCreator.putBytes((byte[]) outputValue);
+              break;
+            default:
+              throw new IllegalStateException();
+          }
+        }
+      } else {
+        for (Object outputValue : outputValues) {
+          switch (fieldSpec.getDataType().getStoredType()) {
+            // Casts are safe here because we've already done the conversion in createDerivedColumnV1Indices
+            case INT:
+              forwardIndexCreator.putIntMV(ArrayUtils.toPrimitive((Integer[]) outputValue));
```

Comment:
Ideally we can directly cast to array of primitive type in the previous step. Let me know if you want to address that in the same PR, or we can add a TODO to address it later

Reply:
We're using the primitive wrapper classes [here](https://github.com/apache/pinot/blob/363a03eea82a66aef4af724625dc714676018aa5/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java#L612) for instance because in the dictionary MV case, we're casting to `Object[]` which isn't possible with primitive arrays. So we could add separate branches in each case [here](https://github.com/apache/pinot/blob/363a03eea82a66aef4af724625dc714676018aa5/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java#L603) for multi-value fields to either create a primitive array (for raw encoded column) or an array of primitive wrapper classes (for dictionary encoded column). But then, I think we'd also need to update all the column statistics collector classes, which currently only handle two cases - the entry is either something that can be casted to an `Object[]` (i.e., an `Integer[]` for `IntColumnPreIndexStatsCollector`), or else its treated as a single-value integer. Primitive arrays (`int[]` for instance) don't seem to be currently handled appropriately (since they can't be cast to `Object[]`) - see [here](https://github.com/apache/pinot/blob/363a03eea82a66aef4af724625dc714676018aa5/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/stats/IntColumnPreIndexStatsCollector.java#L38-L61) for example.

I can certainly make that change here in this PR though if you think it makes sense. Alternatively, I'd be happy to file a follow-up PR to make this optimization and related changes too. Thoughts?

Reply:
Thanks for the detailed explanation! We can solve it separately

Reply:
Yeah, I created a separate PR for it - https://github.com/apache/pinot/pull/13115.